### PR TITLE
Add LFM POC for CT (#19)

### LINKED ISSUE #19: No POC implementation of Latent Flow Matching for CT data
There is currently no implementation of Latent Flow Matching for the dataset of intereset. One should be created in a notebook. These changes should be made:
- Modify `src/data_utils.py` to output a JSON file for VAE.
- Add a notebook which attempts to train a VAE from scratch.
- Generate some CT scans through the use of a foundation model (fine-tuning should be done later).

### PR DESCRIPTION
**Pull request by Couch-bit**
## Description

This PR adds a notebook containing a relatively futile attempt to train a VAE from scratch on available data. In addition, it adds a notebook which generates CTs for test data using weights from a foundation model (this needs future fine-tuning to be able to be properly used).

## Related Issue

Closes #19

## Proposed Changes

- Modifies `src/data_utils.py` to output a JSON file for VAE.
- Adds a notebook which attempts to train a VAE from scratch.
- Adds a notebook which generates CT scans through the use of a foundation model (fine-tuning should be done later).

### CODE REVIEW

File: notebooks/ct_vae_from_scratch.ipynb [hunk 1]
```diff
@@ -0,0 +1,1004 @@
...
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "id": "8345f2fa-7e38-4254-a0c8-7a38f20d4797",
+   "metadata": {},
+   "outputs": [],
+   "source": [
```
Comment (MagicWiqqu) [inline notebooks/ct_vae_from_scratch.ipynb:254]: Personally, given how this is not something as trivial as a classic image recognition, I would leave brief comments explaining why such functions are called in the pipeline. Example -> ScaleIntensityRanged commented as clipping the original HU CT values to 0-1 scale for numerical stability. Either way, I am in favor of explaining the decisions more clearly - it also backs up our decision-making process, even though this approach has been abandoned in favor of fine-tuning the other model due to the hardware limitations.

File: notebooks/ct_vae_from_scratch.ipynb [hunk 2]
```diff
@@ -0,0 +1,1004 @@
...
+    "validation_transform = Compose(\n",
+    "    [\n",
+    "        LoadImaged(keys=\"image\"),\n",
+    "        EnsureChannelFirstd(keys=\"image\"),\n",
+    "        Orientationd(keys=\"image\", axcodes=\"RAS\", labels=None),\n",
+    "        ScaleIntensityRanged(keys=\"image\", a_min=-1000, a_max=1000, b_min=0, b_max=1, clip=True),\n",
+    "        # Pad the image so depth is a multiple of 32\n",
+    "        DivisiblePadd(keys=\"image\", k=32),\n",
+    "        CopyItemsd(keys=[\"image\"], times=1, names=[\"mask\"]),\n",
+    "        ThresholdIntensityd(keys=[\"mask\"], threshold=POS_THRESHOLD, above=True, cval=0),\n",
```
Comment (MagicWiqqu) [inline notebooks/ct_vae_from_scratch.ipynb:280]: Especially this trick needs to be explained. Briefly! But it needs to be documented to be self-explanatory and to back our reasoning.

File: notebooks/ct_vae_from_scratch.ipynb [hunk 3]
```diff
@@ -0,0 +1,1004 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "id": "a2e8697a-a745-4d08-9592-90f60dd4b648",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Not actually equivalent to log(cosh(x)), but is a very good approximation for small x\n",
+    "# and converges to L1 loss for large x\n",
+    "def log_cosh_loss(prediction, target):\n",
```
Comment (MagicWiqqu) [inline notebooks/ct_vae_from_scratch.ipynb:514]: Normally, I'd require a source for this loss (maybe some publication to back up this information), considering our plans for this project (manuscript). Nothing fancy, but to keep track of our sources. But this is an abandoned approach, so OK.

File: notebooks/ct_vae.ipynb [hunk 4]
```diff
@@ -0,0 +1,1004 @@
...
+    "    total_steps = math.floor(epochs * len(train_loader) / accumulation_steps)\n",
+    "\n",
+    "    # Define learning rate scheduler\n",
+    "    scheduler = OneCycleLR(\n",
+    "        optimizer,\n",
+    "        max_lr=lr_max,\n",
+    "        total_steps=total_steps,\n",
+    "        pct_start=lr_warmup,\n",
+    "        anneal_strategy=\"cos\",\n",
+    "        div_factor=lr_start / lr_max,\n",
```
Comment (MagicWiqqu) [inline notebooks/ct_vae.ipynb]: Are those values for `div_factor` and `final_div_factor` intended? According to documentation: 

- div_factor ([float](https://docs.python.org/3/library/functions.html#float)) – Determines the initial learning rate via initial_lr = max_lr/div_factor Default: 25
- final_div_factor ([float](https://docs.python.org/3/library/functions.html#float)) – Determines the minimum learning rate via min_lr = initial_lr/final_div_factor Default: 1e4"

So it is flipped here, and not really commented.

File: notebooks/ct_vae_from_scratch.ipynb (hunk 1 above)
Comment (Couch-bit) [reply notebooks/ct_vae_from_scratch.ipynb:254]: Okay, I can add some extra comments to this notebook when I'm home. Do you think that some extra information is needed for the first "working" pipeline for MAISI notebook, too?

Comment (MagicWiqqu) [reply notebooks/ct_vae_from_scratch.ipynb:254]: In terms of MAISI PoC, if you do notice some parts can be documented more, please do implement it, but I will highlight what seems necessary to me once I proceed with the review. In terms of the pipeline and modularized functions, it should also be included there, along with more detailed documentation.

File: notebooks/maisi_ct_generation.ipynb [hunk 5]
```diff
@@ -0,0 +1,776 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "6742f216-9313-49cc-a93e-a7ae1b92ba69",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# DATA LOCATION CONFIG\n",
```
Comment (MagicWiqqu) [inline notebooks/maisi_ct_generation.ipynb:149]: If something here is rather exotic compared to the standard settings of the MAISI model, please leave a comment as well. From what I see, it's rather straightforward, but you'd know better.

File: notebooks/maisi_ct_generation.ipynb [hunk 6]
```diff
@@ -0,0 +1,776 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "id": "5780384a-cf0c-4529-ab39-48d3759534c5",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# It's possible to center the scans around the prostate since the mask is available, but keeping it simple\n",
```
Comment (MagicWiqqu) [inline notebooks/maisi_ct_generation.ipynb:288]: I wonder if it would be something worth exploring in the future - centering around the prostate could be interesting.

**Review by MagicWiqqu — CHANGES_REQUESTED** [review]
Review my comments, apply changes if you agree with them - ask for clarification and discussion if not.

File: notebooks/ct_vae_from_scratch.ipynb (hunk 3 above)
Comment (Couch-bit) [reply notebooks/ct_vae_from_scratch.ipynb:514]: Yes, I would add a source if it was used in the final approach or if it was very niche.

File: notebooks/ct_vae.ipynb (hunk 4 above)
Comment (Couch-bit) [reply notebooks/ct_vae.ipynb]: This is an error, the ratio is supposed to be other way around like you said. I should probably give this another run after suggestions.

File: notebooks/maisi_ct_generation.ipynb (hunk 5 above)
Comment (Couch-bit) [reply notebooks/maisi_ct_generation.ipynb:149]: Nothing particularly exotic here.

File: notebooks/maisi_ct_generation.ipynb (hunk 6 above)
Comment (Couch-bit) [reply notebooks/maisi_ct_generation.ipynb:288]: Yes, I think some EDA should be done on whether the prostate can deviate from the center and if the center crop actually covers the dose and masks of interest in every sample.

File: notebooks/ct_vae_from_scratch.ipynb [hunk 7]
```diff
@@ -0,0 +1,1035 @@
...
+   "execution_count": 11,
+   "id": "279d6744-c9f9-4360-ae88-e6d688c8df55",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Original shape: (512, 512, 126)\n",
+      "Original affine: [[1.171875 0.       0.       0.      ]\n",
```
Comment (MagicWiqqu) [inline notebooks/ct_vae_from_scratch.ipynb:306]: Is this example CBCT? I wouldn't leave it here as an example of our data as we established it's an outlier. However, if it proves to be messy to re-run it or such, it's ok.

File: notebooks/maisi_ct_generation.ipynb [hunk 8]
```diff
@@ -0,0 +1,873 @@
...
...KuNIRR+eyzz3Zvv/327p577tl95Stf2X31q1/dC+JCQrkL5nzXsRLMhfYnvVJKKaWUUkoppZRyOalgXkq58kgwl1j+7rvv7j755JPdRx99tBfNEchZ0NOFc949zvl99923+4d/+Ie9+O5x0UsppZRSSimllFLK5eAP/unI1enqNVlKudWgXbv99tv3McgldgsXwlfH4WEO8ka/++67d7/97W/3gnkX/iyH6D1SSimllFJKKaVcPCqYl1KuLAje8ib/3e9+d+b08DyvEFqOofdJKaWUUkoppZRy8fjnFexKKeWK4Qt6ngcKw1IRtJRSSimllFJKKeXyUsG8lHKlIbxKKaWUUkoppZRSSikVzEsppZRSSimllFJKKaWUepiXUsrvF+wspZRSSimllFJKKaUe5qWUK03jjpdSSimllFJKKaUUqGBeStld9RjmpZRSSimllFJKKaWICuallN1VF8wrmpdSSimllFJKKaUUUcG8lHKl+ad/+qf9q5RSSimllFJKKaWUCuallCtNxfJSSimllFJKKaWUAhXMSylXnormpZRSSimllFJKKaWCeSnlyqP45RXMSymllFJKKaWUUkoF81LKlacLfpZSSimllFJKKaUU+IN/OtK1sqJSKaWUcn70yYZSSimllFJKKeXi0RjmpZRSSimllFJKKaWUUkoF81JKKaWUUkoppZRSSinl99TDvJRSSimllFJKKaWUUkqpYF5KKaWUUkoppZRSSiml/J56mJdSSimllFJKKaWUUkopFcxLKaWUUkoppZRSSimllN9TD/NSSimllFJKKaWUUkoppYJ5KaWUUkoppZRSSimllPJ76mFeSimllFJKKaWUUkoppVQwL6WUUkoppZRSSimllFJ+Tz3MSymllFJKKaWUUkoppZQK5qWUUkoppZRSSimllFLK7/mj3ZH80z/907G7llJKKaWUUkoppZRSSimXjoZkKaWUUkoppZRSSimllFIqmJdSSimllFJKKaWUUkopv6ce5qWUUkoppZRSSimllFJKBfNSSimllFJKKaWUUkop5ffUw7yUUkoppZRSSimllFJKqWBeSimllFJKKaWUUkoppfyeepiXUkoppZRSSimllFJKKRXMSymllFJKKaWUUkoppZTfUw/zUkoppZRSSimllFJKKbuy2/1/FvOSz4jfK6wAAAAASUVORK5CYII=",
```
Comment (MagicWiqqu) [inline notebooks/maisi_ct_generation.ipynb:330]: Same as in VAE notebook - I do not think CBCT should be an example of our data, but it's acceptable if it's a hassle to change it.

**Review by MagicWiqqu — COMMENTED** [review]
LGTM, but left one comment regarding data visualisation before running the models for both changed notebooks. Optional fix.

File: notebooks/ct_vae_from_scratch.ipynb (hunk 7 above)
Comment (Couch-bit) [reply notebooks/ct_vae_from_scratch.ipynb:306]: It's CT, it may look odd, because it uses the old normalization scheme.

File: notebooks/maisi_ct_generation.ipynb (hunk 8 above)
Comment (Couch-bit) [reply notebooks/maisi_ct_generation.ipynb:330]: Same as above

**Review by MagicWiqqu — APPROVED** [review]
OK
